mmdet/models/layers/conditional_transformer.py

```diff
@@ -0,0 +1,346 @@
+    pos_x = x_embed[:, :, None] / dim_t
+    pos_y = y_embed[:, :, None] / dim_t
+    pos_x = torch.stack((pos_x[:, :, 0::2].sin(), pos_x[:, :, 1::2].cos()), dim=3).flatten(2)
+    pos_y = torch.stack((pos_y[:, :, 0::2].sin(), pos_y[:, :, 1::2].cos()), dim=3).flatten(2)
+    if pos_tensor.size(-1) == 2:
+        pos = torch.cat((pos_y, pos_x), dim=2)
+    elif pos_tensor.size(-1) == 4:
+        w_embed = pos_tensor[:, :, 2] * scale
+        pos_w = w_embed[:, :, None] / dim_t
+        pos_w = torch.stack((pos_w[:, :, 0::2].sin(), pos_w[:, :, 1::2].cos()), dim=3).flatten(2)
+
+        h_embed = pos_tensor[:, :, 3] * scale
+        pos_h = h_embed[:, :, None] / dim_t
+        pos_h = torch.stack((pos_h[:, :, 0::2].sin(), pos_h[:, :, 1::2].cos()), dim=3).flatten(2)
+
+        pos = torch.cat((pos_y, pos_x, pos_w, pos_h), dim=2)
+    else:
+        raise ValueError("Unknown pos_tensor shape(-1):{}".format(pos_tensor.size(-1)))
+    return pos
+
+class ConditionalDetrTransformerDecoder(DetrTransformerDecoder):
+
+    def _init_layers(self):
+        self.layers = ModuleList()
+        for i in range(self.num_layers):
+            self.layers.append(
+                ConditionalDetrTransformerDecoderLayer(**self.layer_cfg[i]))
+        self.embed_dims = self.layers[0].embed_dims  # TODO
+        self.post_norm = build_norm_layer(self.post_norm_cfg,self.embed_dims)[1]
+        #####################################################
+        d_model = 256
+        self.query_scale = MLP(d_model, d_model, d_model, 2)
+        self.ref_point_head = MLP(d_model, d_model, 2, 2)
+        for layer_id in range(self.num_layers - 1):
+            self.layers[layer_id + 1].cross_attn.qpos_proj = None
+        #####################################################
+    def forward(self, query, *args, **kwargs):
+        reference_points_before_sigmoid = self.ref_point_head(kwargs['query_pos'])  # [num_queries, batch_size, 2]
+        reference_points = reference_points_before_sigmoid.sigmoid().transpose(0, 1)
+        intermediate = []
+        for layer_id, layer in enumerate(self.layers):
+            obj_center = reference_points[..., :2].transpose(0, 1)
+            if layer_id == 0:
+                pos_transformation = 1
+            else:
+                pos_transformation = self.query_scale(query)
+            # get sine embedding for the query vector
+            query_sine_embed = gen_sineembed_for_position(obj_center)
+            # apply transformation
+            query_sine_embed = query_sine_embed * pos_transformation
+            kwargs['query_sine_embed'] = query_sine_embed
+            kwargs['is_first'] = (layer_id == 0)
+            query = layer(query, *args, **kwargs)
+            if self.return_intermediate:
+                intermediate.append(self.post_norm(query))
+
+        if self.return_intermediate:
+            return [torch.stack(intermediate), reference_points]
+
+        return [query, reference_points]
```

Comment:
why is here a list?

Reply:
Beacause I do not want to reload the decoder function' forward_transfomer 'in detr.py.and I also have changed    out_dec=out_dec.transpose(1,2) to out_dec[0] = out_dec[0].transpose(1, 2) to supoort this

Reply:
https://github.com/Li-Qingyun/mmdetection/blob/eb3a9d17e72aad1814719c420e13858fd76649a9/mmdet/models/detectors/detr.py#L110

Reply:
I have not thought the modification in 3581a00fab18db7f4e09c01d11bcf559321f5c93 is suitable.
We can have a discussion at our next meeting.